### PR TITLE
rpc, net: Expose connections_onion_only in getnetworkinfo RPC output

### PR DESCRIPTION
**Pull request by hebasto**
This is a non-gui part of adding Tor icon to the GUI.

Split from https://github.com/bitcoin-core/gui/pull/86 as [requested](https://github.com/bitcoin-core/gui/pull/86#discussion_r492689909):
> Also, the RPC and net changes are not gui related, so should go into the main repo

The first approach was in #19926.

### CODE REVIEW

**Comment by hebasto** [comment]
The commits moved from https://github.com/bitcoin-core/gui/pull/86 already have two ACKs:
- from @jnewbery https://github.com/bitcoin-core/gui/pull/86#issuecomment-691905056
> Code review ACK https://github.com/bitcoin-core/gui/commit/6549abcb072d73e0ac91c7992cc5d86eb4064b70 (node changes only. Did not fully review qt changes)

- from @Sjors https://github.com/bitcoin-core/gui/pull/86#pullrequestreview-489554178
> tACK https://github.com/bitcoin-core/gui/commit/6549abcb072d73e0ac91c7992cc5d86eb4064b70 with notes for followup

**Review by jonatack — COMMENTED** [review]
Approach ACK, only skimmed the code so far but the changes look good. Needs test coverage for the new `getnetworkinfo` "connections_onion_only" field. Debug build is clean and I did some quick manual testing. The counts fields contain the correct values. With mixed peers:

```
  "connections_onion_only": false,
```

With only onion peers:
```
  "connections_onion_only": true,
```
RPC help:
```
  "connections_onion_only" : true|false,             (boolean) whether all connection are through the onion network
```

**Comment by Sjors** [comment]
Concept ACK.

Agree that a functional test for `connections_onion_only` would be good, but not sure how easy it is. You could spin up two test nodes, 0 with default settings, 1 with `-onlynet=onion` and `-bind=...=onion`, and then connect node 0 to node 1 via it's "onion" port. In that case node 1 should have `connections_onion_only=true`.

**Comment by hebasto** [comment]
Updated 174848ffb88767fe9190dd7fa160e7187dfcdc07 -> 9371757d9031ad40ca99543bb4d24fc48e00b32e ([pr20172.01](https://github.com/hebasto/bitcoin/commits/pr20172.01) -> [pr20172.02](https://github.com/hebasto/bitcoin/commits/pr20172.02), [diff](https://github.com/hebasto/bitcoin/compare/pr20172.01..pr20172.02)):

- addressed all of @jonatack's and @Sjors' comments
- added a functional test

File: src/rpc/net.cpp [hunk 1]
```diff
@@ -533,10 +533,11 @@ static RPCHelpMan getnetworkinfo()
                         }},
                         {RPCResult::Type::BOOL, "localrelay", "true if transaction relay is requested from peers"},
                         {RPCResult::Type::NUM, "timeoffset", "the time offset"},
+                        {RPCResult::Type::BOOL, "networkactive", "whether p2p networking is enabled"},
                         {RPCResult::Type::NUM, "connections", "the total number of connections"},
                         {RPCResult::Type::NUM, "connections_in", "the number of inbound connections"},
                         {RPCResult::Type::NUM, "connections_out", "the number of outbound connections"},
-                        {RPCResult::Type::BOOL, "networkactive", "whether p2p networking is enabled"},
+                        {RPCResult::Type::BOOL, "connections_onion_only", "whether all connection are through the onion network"},
```
[the remark this answers is not in the document]
Comment (hebasto) [reply src/rpc/net.cpp]: Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/20172#issuecomment-710886459).

File: src/net.h [hunk 2]
```diff
@@ -183,18 +183,28 @@ enum class ConnectionType {
...
+    const bool onion_only{false};
+
+    ConnCounts()
+    {
+    }
+
+    ConnCounts(int num_in, int num_out, bool is_onion_only)
+        : all(num_in + num_out), in(num_in), out(num_out), onion_only(is_onion_only)
+    {
+    }
```
[the remark this answers is not in the document]
Comment (hebasto) [reply src/net.h]: Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/20172#issuecomment-710886459).

**Comment by hebasto** [comment]
Hmm, the current [pr20172.03](https://github.com/hebasto/bitcoin/commits/pr20172.03) branch does not pass macOS 10.14 build on Travis.

The alternative [pr20172.04](https://github.com/hebasto/bitcoin/commits/pr20172.04) branch pass locally on macOS 10.15 and Linux Mint 20 (Focal codebase), but it does not pass on Travis at all.

What is the correct portable way to `-bind=localhost` and `-bind=localhost:port=onion` simultaneously?

**Comment by hebasto** [comment]
Updated 9371757d9031ad40ca99543bb4d24fc48e00b32e -> d301cc5c564f8c130949567602e182917efef18e ([pr20172.02](https://github.com/hebasto/bitcoin/commits/pr20172.02) -> [pr20172.06](https://github.com/hebasto/bitcoin/commits/pr20172.06), [diff](https://github.com/hebasto/bitcoin/compare/pr20172.02..pr20172.06)):

- fixed functional test

Travis is green now!

File: test/functional/feature_onion.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        self.log.info("Accepting an inbound 'onion'-tagged connection, and checking...")
+        self.nodes[1].addnode(self.bind_onion, "onetry")
+        network_info = self.nodes[0].getnetworkinfo()
+        assert_equal(network_info["connections"], 1)
+        assert_equal(network_info["connections_in"], 1)
+        assert_equal(network_info["connections_out"], 0)
+        assert_equal(network_info["connections_onion_only"], True)
+
+        self.log.info("Creating an outbound onion connection, and checking...")
+        self.nodes[0].addnode("bitcoinostk4e4re.onion:8333", "onetry")
```
Comment (jnewbery) [inline test/functional/feature_onion.py:61]: What is this onion address?

**Review by jnewbery — COMMENTED** [review]
One question and one small naming suggestion.

This PR contains qt-only changes. It shouldn't be too difficult to separate the changes so that PRs to the main repo only contain changes to the node code, and PRs to the GUI repo only contain changes to qt code.

File: test/functional/feature_onion.py (hunk 3 above)
Comment (hebasto) [reply test/functional/feature_onion.py:61]: It has been taken from https://github.com/bitcoin/bitcoin/blob/d9d9a293520d26ca8fec8caebab46db103b29903/test/functional/feature_proxy.py#L139

**Comment by hebasto** [comment]
@jnewbery 
> This PR contains qt-only changes. It shouldn't be too difficult to separate the changes so that PRs to the main repo only contain changes to the node code, and PRs to the GUI repo only contain changes to qt code.

Clear qt-only changes are already [leaved](https://github.com/bitcoin-core/gui/pull/86) in the GUI repo. This pull contains two commits that modify signal signatures. As these signals propagate through the qt code, I do not want to break those commits in "non-qt" and "qt" parts.

File: test/functional/feature_onion.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        self.add_nodes(self.num_nodes, extra_args=args)
+        self.start_nodes()
+
+    def run_test(self):
+        self.log.info("Checking a node without any peers...")
+        network_info = self.nodes[0].getnetworkinfo()
+        assert_equal(network_info["connections"], 0)
+        assert_equal(network_info["connections_in"], 0)
+        assert_equal(network_info["connections_out"], 0)
+        assert_equal(network_info["connections_onion_only"], False)
```
Comment (jnewbery) [inline test/functional/feature_onion.py:50]: This repetitious code could potentially be refactored into a small helper function `get_connection_info()`.

File: test/functional/feature_onion.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        assert_equal(network_info["connections_in"], 1)
+        assert_equal(network_info["connections_out"], 0)
+        assert_equal(network_info["connections_onion_only"], True)
+
+        self.log.info("Creating an outbound onion connection, and checking...")
+        self.nodes[0].addnode("bitcoinostk4e4re.onion:8333", "onetry")
+        network_info = self.nodes[0].getnetworkinfo()
+        assert_equal(network_info["connections"], 2)
+        assert_equal(network_info["connections_in"], 1)
+        assert_equal(network_info["connections_out"], 1)
```
Comment (jnewbery) [inline test/functional/feature_onion.py:65]: I'm confused about how this works. It creates an outbound connection to where? Why does that outbound connection show up here, but is gone by the time you call `getnetworkinfo()` again?

Comment (hebasto) [reply test/functional/feature_onion.py:65]: > It creates an outbound connection to where?

To created `self.proxy`.

> Why does that outbound connection show up here, but is gone by the time you call `getnetworkinfo()` again?

IDK why `add_p2p_connection(P2PInterface())` causes drop of the connection to the proxy.

Comment (hebasto) [reply test/functional/feature_onion.py]: > Why does that outbound connection show up here, but is gone by the time you call `getnetworkinfo()` again?

Connections to a dummy Socks5 server are short-lived:https://github.com/bitcoin/bitcoin/blob/45385018e13e167521e655c36128d8ee4f2a3e0b/test/functional/test_framework/socks5.py#L113-L124

**Comment by hebasto** [comment]
Updated d301cc5c564f8c130949567602e182917efef18e -> caebabfc4b239bfb854228f041a6b4220e87e3af ([pr20172.06](https://github.com/hebasto/bitcoin/commits/pr20172.06) -> [pr20172.09](https://github.com/hebasto/bitcoin/commits/pr20172.09), [diff](https://github.com/hebasto/bitcoin/compare/pr20172.06..pr20172.09)):

- reworked functional test

- addressed @jnewbery's [comment](https://github.com/bitcoin/bitcoin/pull/20172#discussion_r507626814):
> This repetitious code could potentially be refactored into a small helper function `get_connection_info()`.

File: src/net.h [hunk 6]
```diff
@@ -346,7 +350,8 @@ class CConnman
     bool RemoveAddedNode(const std::string& node);
     std::vector<AddedNodeInfo> GetAddedNodeInfo();
 
-    size_t GetNodeCount(NumConnections num);
+    size_t PeerCount() { return WITH_LOCK(cs_vNodes, return vNodes.size()); }
+    ConnCounts ConnectionCounts();
```
[the remark this answers is not in the document]
Comment (hebasto) [reply src/net.h]: Sure! [Updated](https://github.com/bitcoin/bitcoin/pull/20172#issuecomment-712479264).

File: test/functional/feature_onion.py (hunk 4 above)
Comment (hebasto) [reply test/functional/feature_onion.py]: Thanks! [Reworked](https://github.com/bitcoin/bitcoin/pull/20172#issuecomment-712479264).

File: test/functional/feature_onion.py (hunk 5 above)
Comment (hebasto) [reply test/functional/feature_onion.py]: [Reworked](https://github.com/bitcoin/bitcoin/pull/20172#issuecomment-712479264).

File: test/functional/feature_onion.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        self.num_nodes = 2
+        self.setup_clean_chain = True
+
+    def setup_network(self):
+        self.add_nodes(self.num_nodes)
+        self.start_nodes()
+
+    def test_connections(self, in_clearnet=0, in_onion=0, out_onion=0):
+        self.restart_node(0, ['-bind={}=onion'.format(self.bind_onion), self.arg_onion])
+        self.restart_node(1)
```
Comment (jnewbery) [inline test/functional/feature_onion.py:33]: Stop-starting the nodes 11 times just to test new connections seems very inefficient. Is there a reason you need to do this stop-start?

**Comment by hebasto** [comment]
Updated 85ff09ef5d4da9991ee1e9761662d003ff464679 -> da8b1092fc2d5fa5ee9da3310dae16d4c919bc1b ([pr20172.10](https://github.com/hebasto/bitcoin/commits/pr20172.10) -> [pr20172.11](https://github.com/hebasto/bitcoin/commits/pr20172.11), [diff](https://github.com/hebasto/bitcoin/compare/pr20172.10..pr20172.11)):

- addressed @jnewbery's [comment](https://github.com/bitcoin/bitcoin/pull/20172#discussion_r508290004):
> Stop-starting the nodes 11 times just to test new connections seems very inefficient. Is there a reason you need to do this stop-start?

- improved robustness

File: test/functional/feature_onion.py (hunk 7 above)
Comment (hebasto) [reply test/functional/feature_onion.py]: Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/20172#issuecomment-712763616).

**Comment by Sjors** [comment]
tACK da8b109

**Review by luke-jr — CHANGES_REQUESTED** [review]
Concept NACK per https://github.com/bitcoin-core/gui/pull/86#pullrequestreview-516300696

**Comment by kristapsk** [comment]
I'm thinking about this in a context of if / when other privacy network support is added to Bitcoin Core (like I2P, #20254). What people might want then is not "are all of my connections Tor only?", but "are all of my connections using privacy networks?" and "what kind of privacy networks are they using?".

**Comment by hebasto** [comment]
Closing. Leaving it up for grabs.

**Pull request closed without merging**